File: docs/js/blackjack/bankroll.js
export function settleHand(playerHand, dealerHand, dealerHasBlackjack) {
  const bet = playerHand.bet;

  if (playerHand.isSurrendered) {
    return { description: "SURRENDER", netChange: -(bet >> 1) };
  }
  if (playerHand.isBust) {
    return { description: "BUST", netChange: -bet };
  }

  const playerBj = playerHand.isBlackjack;
  if (playerBj && dealerHasBlackjack) return { description: "PUSH (both blackjack)", netChange: 0 };
  if (playerBj) return { description: "BLACKJACK", netChange: Math.floor((bet * 3) / 2) };
  if (dealerHasBlackjack) return { description: "LOSE (dealer blackjack)", netChange: -bet };
  if (dealerHand.isBust) return { description: "WIN (dealer bust)", netChange: bet };
  if (playerHand.value > dealerHand.value) return { description: "WIN", netChange: bet };
  if (playerHand.value < dealerHand.value) return { description: "LOSE", netChange: -bet };
  return { description: "PUSH", netChange: 0 };
}

export function settleInsurance(insuranceBet, dealerHasBlackjack) {
  if (insuranceBet <= 0) return 0;
  if (dealerHasBlackjack) return insuranceBet * 2;
  return -insuranceBet;
}

export function computeBotBet(bankroll, config) {
  if (bankroll < config.minBet) return 0;
  let target = Math.max(config.minBet, Math.floor(bankroll * 0.03));
  let rounded = Math.floor(target / config.minBet) * config.minBet;
  if (rounded < config.minBet) rounded = config.minBet;
  return Math.max(config.minBet, Math.min(rounded, config.maxBet, bankroll));
}

File: blackjack/bankroll.py
from __future__ import annotations

from dataclasses import dataclass

from blackjack.config import GameConfig
from blackjack.hand import Hand


def clamp_bet(amount: int, bankroll: int, config: GameConfig) -> int:
    return max(config.min_bet, min(amount, config.max_bet, bankroll))


def compute_bot_bet(bankroll: int, config: GameConfig) -> int:
    if bankroll < config.min_bet:
        return 0
    target = max(config.min_bet, int(bankroll * 0.03))
    rounded = (target // config.min_bet) * config.min_bet
    if rounded < config.min_bet:
        rounded = config.min_bet
    return clamp_bet(rounded, bankroll, config)


@dataclass
class SettlementResult:
    description: str
    net_change: int


def settle_hand(
    player_hand: Hand,
    dealer_hand: Hand,
    *,
    dealer_has_blackjack: bool,
    blackjack_payout_numerator: int = 3,
    blackjack_payout_denominator: int = 2,
) -> SettlementResult:
    bet = player_hand.bet

    if player_hand.is_surrendered:
        loss = bet // 2
        return SettlementResult("SURRENDER", -loss)

    if player_hand.is_bust:
        return SettlementResult("BUST", -bet)

    player_bj = player_hand.is_blackjack
    if player_bj and dealer_has_blackjack:
        return SettlementResult("PUSH (both blackjack)", 0)
    if player_bj:
        win = bet * blackjack_payout_numerator // blackjack_payout_denominator
        return SettlementResult("BLACKJACK", win)
    if dealer_has_blackjack:
        return SettlementResult("LOSE (dealer blackjack)", -bet)
    if dealer_hand.is_bust:
        return SettlementResult("WIN (dealer bust)", bet)
    if player_hand.value > dealer_hand.value:
        return SettlementResult("WIN", bet)
    if player_hand.value < dealer_hand.value:
        return SettlementResult("LOSE", -bet)
    return SettlementResult("PUSH", 0)


def settle_insurance(insurance_bet: int, dealer_has_blackjack: bool) -> int:
    if insurance_bet <= 0:
        return 0
    if dealer_has_blackjack:
        return insurance_bet * 2
    return -insurance_bet
